File: 28/28.cpp
#include<iostream>
using namespace std;

class Solution {
public:
    int next[50010],n,m;
    void getNext(string str) {
        int i = -1,j = 0;
        next[0] = -1;
        while(j < m) {
            if(i == -1 or str[i] == str[j]) {
                i++,j++;
                next[j] = i;
            } else {
                i = next[i];
            }
        }
    }
    int strStr(string haystack, string needle) {
        n = haystack.size(),m = needle.size();
        if(!m)  return 0;

        getNext(needle);
        // kmp
        int i = 0,j = 0;
        while(j < n) {
            if(i == -1 or needle[i] == haystack[j]) {
                i++,j++;
            } else {
                i = next[i];
            }
            if (i == m) {
                return j-i;
            }
        }
        return -1;
    }
};

int main () {
    Solution a;
    cout << a.strStr("hello", "ll") << endl;
}

File: 28/28.go
package main

import "fmt"

func strStr(haystack string, needle string) int {
	if needle == "" {
		return 0
	}
	if haystack == "" {
		return -1
	}
	// get next
	next := make([]int, len(needle)+1)
	i, j := -1, 0
	next[0] = -1
	for j < len(needle) {
		if i == -1 || needle[i] == needle[j] {
			i, j = i+1, j+1
			next[j] = i
		} else {
			i = next[i]
		}
	}
	// kmp
	i, j = 0, 0
	for j < len(haystack) {
		if i == -1 || needle[i] == haystack[j] {
			i, j = i+1, j+1
		} else {
			i = next[i]
		}
		if i == len(needle) {
			return j - i
		}
	}
	return -1
}

func main() {
	fmt.Println(strStr("aa", "aa"))
}
